File: ekconfig/ekconf_dir/menu.cc
/*
* Ekconf, an Eclipe plug-in for configuring the Linux kernel or Buildroot.
* 
* Copyright (C) 2012 Tiana Rakotovao <rakotovaomahefa@gmail.com>
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 52 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <iostream>
#include "menu.h"
using std::cout;

jlong getpList(JNIEnv *env ,jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu *) menuF;
	return (jlong) pt->list;
}

jlong getpNext(JNIEnv *env, jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu* ) menuF;
	return (jlong) pt->next;
}

jlong getpSym(JNIEnv *env , jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu* ) menuF;
	return (jlong) pt->sym;
}

jboolean hasProperty(JNIEnv *env , jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu* ) menuF;
	if (pt->prompt)
		return JNI_TRUE;
	else
		return JNI_FALSE;
}

jstring getPrompt(JNIEnv *env, jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu *) menuF;
	if (pt->prompt)
		return JNU_NewStringNative(env,menu_get_prompt(pt)) ;
	else
		return NULL;
}

jboolean isVisible(JNIEnv *env , jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu *) menuF;
	if (menu_is_visible(pt))
		return JNI_TRUE;
	return JNI_FALSE;
}

jstring getHelp(JNIEnv *env , jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	struct gstr help;
	jstring result;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu *) menuF;
	help = str_new();
	menu_get_ext_help(pt, &help);
	result = JNU_NewStringNative(env , str_get(&help));
	str_free(&help);
	return result;
}

jint getPropertyType(JNIEnv *env , jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu *) menuF;
	if(pt->prompt)
		return (jint) pt->prompt->type;
	else
		return (jint)-1;
}

jlong getpParent(JNIEnv *env , jobject jmenu)
{
	jlong menuF;
	struct menu *pt;
	menuF = env->GetLongField(jmenu , FID_Menu_pMenu);
	JNU_CheckAndThrowException(env);
	pt = (struct menu *) menuF;
	return (jlong) pt->parent;
}

File: plugin/src/ekconf/kconfig/Menu.java
/*
* Ekconf, an Eclipe plug-in for configuring the Linux kernel or Buildroot.
* 
* Copyright (C) 2012 Tiana Rakotovao Andriamahefa <rkmahefa@gmail.com>
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 52 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

package ekconf.kconfig;

import java.util.List;

/**
 * A menu is an entry in a kbuild configuration file.
 * 
 * @author Tiana Rakotovao
 */
public abstract class Menu {
	
	private Menu parent = null;
	private List<Menu> children = null;
	private Symbol symbol = null;
	private String prompt = null;
	private PropertyType propertyType = null;

	protected Menu(){}
	
	public Menu getParent() {
		return parent;
	}
	
	protected void setParent(Menu parent) {
		this.parent = parent;
	}
	
	protected void setChildren(List<Menu> children)
	{
		this.children = children;
	}
	
	public List<Menu> getChildren(){
		return children;
	}

	public boolean hasChildren()
	{
		if (null == this.children)
			return false;
		return true;
	}
	
	public abstract boolean hasSibling();
	
	public boolean hasSymbol() {
		if (null == this.symbol)
			return false;
		return true;
	}
	
	@Override
	public String toString(){
		if (this.prompt != null)
			return this.prompt;
		else {
			if (this.hasSymbol())
				return this.getSymbol().getName();
		}
		return null;
	}

	public String getPrompt() {
		return prompt;
	}

	protected void setPrompt(String prompt) {
		this.prompt = prompt;
	}
	
	public boolean hasPrompt() {
		return (null != this.prompt);
	}
	
	public Symbol getSymbol(){
		return symbol;
	}
	
	protected void setSymbol(Symbol symbol) {
		this.symbol = symbol;
	}
	
	public abstract boolean hasProperty();
	
	public PropertyType getPropertyType() {
		return this.propertyType;
	}
	
	protected void setPropertyType(PropertyType prop)
	{
		this.propertyType = prop;
	}
	
	public abstract boolean isVisible();
	
	public abstract String getHelp();

}
